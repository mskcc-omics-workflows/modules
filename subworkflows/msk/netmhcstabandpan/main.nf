include { NEOANTIGENUTILS_GENERATEHLASTRING  } from '../../../modules/msk/neoantigenutils/generatehlastring/main'
include { NEOANTIGENUTILS_GENERATEMUTFASTA  } from '../../../modules/msk/neoantigenutils/generatemutfasta/main'
include { NETMHCPAN4 } from '../../../modules/msk/netmhcpan4/main'
include { NETMHC3 } from '../../../modules/msk/netmhc3/main'
include { NETMHCSTABPAN } from '../../../modules/msk/netmhcstabpan/main'
include { NEOANTIGENUTILS_FORMATNETMHCPAN } from '../../../modules/msk/neoantigenutils/formatnetmhcpan/main'

workflow NETMHCSTABANDPAN {

    take:

    ch_maf_and_hla         // channel: [ val(meta), maf, hla ]
    ch_cds_and_cdna        // channel: [ cfs, cdna]
    ch_neosv_out

    main:

    ch_versions = Channel.empty()

    ch_hla = ch_maf_and_hla
                .map{
                    new Tuple(it[0],it[2])
                }


    ch_maf = ch_maf_and_hla
                .map{
                    new Tuple(it[0],it[1])
                }


    NEOANTIGENUTILS_GENERATEHLASTRING( ch_hla )


    ch_versions = ch_versions.mix(NEOANTIGENUTILS_GENERATEHLASTRING.out.versions)

    NEOANTIGENUTILS_GENERATEMUTFASTA( ch_maf, ch_cds_and_cdna )

    ch_versions = ch_versions.mix(NEOANTIGENUTILS_GENERATEMUTFASTA.out.versions)

    ch_netmhcinput = createNETMHCInput(NEOANTIGENUTILS_GENERATEMUTFASTA.out.wt_fasta,
                                        NEOANTIGENUTILS_GENERATEMUTFASTA.out.mut_fasta,
                                        NEOANTIGENUTILS_GENERATEHLASTRING.out.hlastring,
                                        ch_neosv_out
                                        )
    
    NETMHCSTABPAN( ch_netmhcinput )

    ch_versions = ch_versions.mix(NETMHCSTABPAN.out.versions)

    merged_pan_and_stab = Channel.empty()

    if ( params.netmhc3 ) {
        
        NETMHC3( ch_netmhcinput )
        ch_versions = ch_versions.mix(NETMHC3.out.versions)
        merged_pan_and_stab = NETMHC3.out.netmhcoutput.mix(NETMHCSTABPAN.out.netmhcstabpanoutput)
    }
    else{

        NETMHCPAN4( ch_netmhcinput )
        ch_versions = ch_versions.mix(NETMHCPAN4.out.versions)
        merged_pan_and_stab = NETMHCPAN4.out.netmhcpanoutput.mix(NETMHCSTABPAN.out.netmhcstabpanoutput)
    }    

    NEOANTIGENUTILS_FORMATNETMHCPAN( merged_pan_and_stab )

    ch_versions = ch_versions.mix( NEOANTIGENUTILS_FORMATNETMHCPAN.out.versions )

    emit:

    tsv        = NEOANTIGENUTILS_FORMATNETMHCPAN.out.netMHCpanreformatted     // channel: [ val(meta), [ tsv ] ]
    //xls        = NETMHCPAN.out.xls                                            // channel: [ val(meta), [ xls ] ]
    mut_fasta  = NEOANTIGENUTILS_GENERATEMUTFASTA.out.mut_fasta               // channel: [ val(meta), [ *.MUT_sequences.fa ] ]
    wt_fasta   = NEOANTIGENUTILS_GENERATEMUTFASTA.out.wt_fasta                // channel: [ val(meta), [ *.WT_sequences.fa ] ]
    versions   = ch_versions                                                  // channel: [ versions.yml ]
}

def createNETMHCInput(wt_fasta, mut_fasta, hla, sv_fastas) {
        mut_fasta_channel = mut_fasta
            .map{
                new Tuple(it[0],it)
                }

        wt_fasta_channel = wt_fasta
            .map{
                new Tuple(it[0],it)
                }

        mut_SVfasta_channel = sv_fastas
            .map{
                new Tuple(it[0],it[1])
                }

        wt_SVfasta_channel = sv_fastas
            .map{
                new Tuple(it[0],it[2])
                }

        hla_channel = hla
            .map{
                new Tuple(it[0],it[1])
                }

        merged_mut_fasta = mut_fasta_channel
            .join(mut_SVfasta_channel, by:0)

        merged_mut = merged_mut_fasta.join(hla_channel)
            .map{
                new Tuple(it[1][0], it[1][1], it[2], it[3],"MUT")
            }

        merged_wt_fasta = wt_fasta_channel
            .join(wt_SVfasta_channel,by: 0)

        merged_wt = merged_mut_fasta
            .join(hla_channel)
            .map{
                new Tuple(it[1][0], it[1][1], it[2], it[3],"WT")
            }
        merged = merged_mut.mix(merged_wt)
        return merged
}
