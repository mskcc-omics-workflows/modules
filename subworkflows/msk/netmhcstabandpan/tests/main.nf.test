nextflow_workflow {

    name "Test Subworkflow NETMHCSTABANDPAN"
    script "../main.nf"
    workflow "NETMHCSTABANDPAN"

    tag "subworkflows"
    tag "subworkflows_nfcore"
    tag "subworkflows_msk"
    tag "subworkflows/netmhcstabandpan"
    tag "netmhcstabandpan"
    tag "neoantigenutils/generatehlastring"
    tag "neoantigenutils/generatemutfasta"
    tag "netmhcpan"
    tag "netmhcstabpan"
    tag "neoantigenutils/formatnetmhcpan"

    test("netmhcstabandpan - tsv,xls,fa") {

        when {
            workflow {
                """
                input[0] = Channel.value([
                    [ id:'test', single_end:false ], // meta map
                    file(params.test_data_mskcc['neoantigen']['temp_test_maf'], checkIfExists: true),
                    file(params.test_data_mskcc['neoantigen']['winners_hla_txt'], checkIfExists: true),
                    ])
                input[1] = Channel.value([
                    file(params.test_data_mskcc['neoantigen']['cds'], checkIfExists: true),
                    file(params.test_data_mskcc['neoantigen']['cdna'], checkIfExists: true)
                    ])

                """
            }
        }

        then {
            assertAll(
                { assert workflow.success},
                { assert snapshot(workflow.out.tsv[0][1],
                                    file(workflow.out.xls[0][1]).name,
                                    workflow.out.mut_fasta[0][1],
                                    workflow.out.wt_fasta[0][1]
                                ).match()
                }
            )
        }
    }

    test("netmhcstabandpan - tsv,xls,fa - stub") {

        options "-stub"

        when {
            workflow {
                """
                input[0] = Channel.value([
                    [ id:'test', single_end:false ], // meta map
                    file('temp_test_maf'),
                    file('winners_hla_txt'),
                    ])
                input[1] = Channel.value([
                    file('cds'),
                    file('cdna')
                    ])
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success},
                { assert snapshot(workflow.out.tsv[0][1],
                                  file(workflow.out.xls[0][1]).name,
                                  workflow.out.mut_fasta[0][1],
                                  workflow.out.wt_fasta[0][1]
                                ).match()
                }

            )
        }
    }
}
