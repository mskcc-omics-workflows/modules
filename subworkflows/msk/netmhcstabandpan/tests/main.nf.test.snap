{
    "netmhcstabandpan - tsv,xls,fa": {
        "content": [
            "test_netmhcpan.output.WT.tsv",
            "test.MUT_sequences.fa:md5,7fdb7d3f0fe5a6f439ed294b612c2d70",
            "test.WT_sequences.fa:md5,7595ed6cf0c98500b00c9ad027125b38"
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.4"
        },
        "timestamp": "2024-12-17T15:32:48.122038795"
    },
    "netmhcstabandpan - tsv,xls,fa - stub": {
        "content": [
            "h",
            "test.MUT_sequences.fa:md5,d41d8cd98f00b204e9800998ecf8427e",
            "test.WT_sequences.fa:md5,d41d8cd98f00b204e9800998ecf8427e"
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.4"
        },
        "timestamp": "2024-12-17T15:33:59.951435808"
    },
    "netmhcstabandnetmhc3 - SV - tsv,xls,fa": {
        "content": [
            "test_netmhc.output.WT.tsv",
            "test.MUT_sequences.fa:md5,118b48df96c7217675b9f9ac14309a25",
            "test.WT_sequences.fa:md5,bb7dfff23ae47cf64ec4854ee48ec78d"
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.4"
        },
        "timestamp": "2024-12-17T15:31:41.007610492"
    },
    "netmhcstabandpan - SV - tsv,xls,fa": {
        "content": [
            "test_netmhcpan.output.WT.tsv",
            "test.MUT_sequences.fa:md5,118b48df96c7217675b9f9ac14309a25",
            "test.WT_sequences.fa:md5,bb7dfff23ae47cf64ec4854ee48ec78d"
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.4"
        },
        "timestamp": "2024-12-17T15:29:06.330489514"
    },
    "netmhcstabandnetmhc3 - tsv,xls,fa": {
        "content": [
            "test_netmhc.output.WT.tsv",
            "test.MUT_sequences.fa:md5,7fdb7d3f0fe5a6f439ed294b612c2d70",
            "test.WT_sequences.fa:md5,7595ed6cf0c98500b00c9ad027125b38"
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.4"
        },
        "timestamp": "2024-12-17T15:33:45.014740679"
    }
}